include/xgboost/tree_updater.h

```diff
@@ -49,18 +49,25 @@ class TreeUpdater : public Configurable {
    *  existing trees.
    */
   virtual bool CanModifyTree() const { return false; }
+  /*!
+   * \brief Wether the out_position in `Update` is valid. This determines whether adaptive
+   *        tree can be used.
+   */
+  virtual bool HasNodePosition() const { return false; }
```

Comment:
In theory we can make every updater support node position right? In xgboost we are using a kind of 'fake object-oriented programming'. We define an interface of functionality that all tree updaters should implement, but then make some updaters implement this functionality and others not.

Reply:
@RAMitchell We have exact updater and pruner.  Surely, we can work on them too, but that's just too much for 1 PR I think.